pkgs/tools/system/lanzaboote/tool.nix

```diff
@@ -0,0 +1,67 @@
+{ systemd
+, makeWrapper
+, stdenv
+, binutils-unwrapped
+, sbsigntool
+, rustPlatform
+, fetchFromGitHub
+, lib
+}:
+let
+  uefiPkgs = import ../../../../. {
```

Comment:
No, because we were not able to get an agreement of what is (x86_64|aarch64|...)-uefi in nixpkgs I believe.

I argue we should try to go with this thing for now and I am interested to know how we can do this better in the future, rust-hypervisor-firmware has the same issue.

Reply:
I really don't want to set even more of a precedent that re-importing Nixpkgs like this is okay.  (I would have objected to the rust-hypervisor-firmware change if I'd noticed it in time.)  We need a consensus on how to handle UEFI.  There are still open questions that are waiting for an answer from you — does building for Windows even work for aarch64 UEFI, for example?  @amjoseph-nixpkgs had a convincing sounding opinion that it would not.

Reply:
> I really don't want to set even more of a precedent that re-importing Nixpkgs like this is okay

:+1: :+1: :+1: :+1:  yes please don't reimport nixpkgs.  That is the path to madness.

It looks like you don't use `uefiPkgs` for anything other than the `hostPlatform` of `buildRustPackage`.  So really all that's needed is a uefi-targeted `rustc`, not a full nixpkgs, right?

If so, we should be able to sidestep the whole debate over what nixpkgs' uefi triple is, because we don't really need one.  We just need a rustc, not a complete packageset.

Reply:
Isn't the question about reimporting nixpkgs at least for x86_64 Linux
orthogonal from aarch64 ?

But yes, I agree. I will check for AArch64 and I will report.

Reply:
Also, I agree we should not reimport nixpkgs but I opened this also so we can discuss it here. :)

Reply:
I will answer the question on AArch64, here's my method:
- Compile a lanzaboote stub for AArch64 via cross-compilation, put it in an ESP directory in the EFI fallback directory or whatever (you can run it via EFI shell).
- Run `export OVMF="$(nix-build -A pkgsCross.aarch64-multiplatform.OVMF.fd)/" ; qemu-system-aarch64 -cpu max -drive if=pflash,format=raw,file="/tmp/QEMU_EFI.raw" -drive format=raw,file=fat:rw:/tmp/esp -net none -nographic -serial mon:stdio -machine virt -m 2G` where `/tmp/esp` is your ESP directory.
- Boot the lanzaboote stub.

It will fail with "did you run lzbt?" which is expected, proving its ability to run code properly on AArch64 if I'm not wrong.

Therefore, it does work @alyssais without any issue on this target, at least, in a VM. For an actual machine, I still need a new systemd release, otherwise I'd need to go through some serious hoops which I don't think are blocker for the "reimport nixpkgs" discussion (are they?).

Reply:
> It looks like you don't use uefiPkgs for anything other than the hostPlatform of buildRustPackage. So really all that's needed is a uefi-targeted rustc, not a full nixpkgs, right?

Yes, I only need a `rustc` able to cross-compile code to a UEFI target.

> If so, we should be able to sidestep the whole debate over what nixpkgs' uefi triple is, because we don't really need one. We just need a rustc, not a complete packageset.

Do you mean to introduce cross compilation logic without reusing `crossSystem` and a different `stdenv` ?

Reply:
BTW, I built the stub with `NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1 nix-build -E '(import <nixpkgs> { crossSystem = { config = "aarch64-windows"; libc = null; rustc.config = "aarch64-unknown-uefi"; useLLVM = true; }; }).lanzaboote-stub'`, I will add in this PR the support for `aarch64-windows` in compiler-rt, given that it's working.

Reply:
Introduced in d212bb81dcd4fa8b724d93e997aff287aa9c3c69

Reply:
> Therefore, it does work @alyssais without any issue on this target, at least, in a VM.

Interesting.  Thoughts, @amjoseph-nixpkgs?

Reply:
I think it doesn't mean much unless you try it on TCG-certified aarch64-uefi hardware.  OVMF specifically states that they [are not UEFI-compliant](https://github.com/tianocore/tianocore.github.io/wiki/OVMF-FAQ#are-ovmf-releases-fully-uefi-compliant).  For all we know the OVMF/qemu developers don't have access to that hardware either so they just had it default to "do the same thing x86 does, but with a different `--target=` flag".  On aarch64 the ATF loads even before any of the UEFI stuff, and they call back and forth -- it's really quite weird, like a corporate power struggle *writ firmware*.  Anyways I don't see `arm-trusted-firmware` anywhere in the derivation-closure of `nix-instantiate -A {qemu,OVMF}`, so I think they just left this stuff out.

This shouldn't be a blocker though.  Just add a comment next to the aarch64 triple that says "`# untested, might be wrong, may change`" and come back to it later when you have aarch64 TCG hardware to test on.

Reply:
Ugh, OK.

So do I gather correctly that the way to resolve this thread is:

(1) comment for aarch64 TCG stuff
(2) have an UEFI `buildRustPackage` rather than re-importing Nixpkgs

Reply:
I'd be curious about the thoughts of @samueldr and @Ericson2314 on the matter, given their familiarity with arm64 and cross compilation triples respectively.

Reply:
I'm trying to do this `UEFI` rustPlatform mechanism but I will need some help as I clearly don't have the required extensive experience in how splicing/stdenv modification works to figure out what I am missing to kickstart the build here. I pushed what I have done.

Reply:
As of now, this is solved in the latest revision of this PR, I am resolving the discussion.